nomic/aws/sagemaker.py

```diff
@@ -0,0 +1,181 @@
+
+NULL_PLACEHOLDER = hashlib.md5(b"nomic null").hexdigest()
+EMPTY_PLACEHOLDER = hashlib.md5(b"nomic empty").hexdigest()
+
+
+class NomicTextEmbeddingModel(Enum):
+    nomic_embed_text_v1_5 = "nomic-embed-text-v1.5"
+
+    def hamming_capable(self):
+        return self == NomicTextEmbeddingModel.nomic_embed_text_v1_5
+
+    def matryoshka_capable(self):
+        return self == NomicTextEmbeddingModel.nomic_embed_text_v1_5
+
+    def dim(self):
+        return text_embedding_model_info[self.value]["dim"]
+
+    def recommended_dims(self):
+        return text_embedding_model_info[self.value].get("recommended_dims") or [
+            self.dim()
+        ]
+
+    def max_length(self):
+        return text_embedding_model_info[self.value]["max_length"]
+
+    def pad_token(self) -> int:
+        return text_embedding_model_info[self.value]["pad_id"]
+
+
+def null_empty_placeholder(text: Optional[str]) -> str:
+    """
+    If text is null or empty replace with a placeholder nonsense string.
+    """
+    if text is None:
+        return NULL_PLACEHOLDER
+    if text.strip() == "":
+        return EMPTY_PLACEHOLDER
+    return text
+
+
+def load_tokenizer(model: NomicTextEmbeddingModel) -> Tokenizer:
+    """
+    Instantiate the huggingface tokenizer.
+    :param model: the model doing the embedding
+    :return:
+    """
+    tokdir = Path(__file__).parent / "tokenizers"
+    if model == NomicTextEmbeddingModel.nomic_embed_text_v1_5:
+        tokenizer = Tokenizer.from_file(str(tokdir / "bert-base-uncased.json"))
+        # TODO: change to 8192
+        tokenizer.enable_truncation(max_length=2048)
+        tokenizer.enable_padding(pad_id=0, pad_token="[PAD]", pad_to_multiple_of=64)
+    else:
+        raise Exception(f"Could not determine tokenizer for model: {model.value}")
+    return tokenizer
+
+
+@contextmanager
+def no_pad(tok):
+    if tok.padding is not None:
+        keep = tok.padding
+        tok.no_padding()
+        yield tok
+        tok.enable_padding(**keep)
+    else:
+        yield tok
+
+
+@contextmanager
+def no_trunc(tok):
+    if tok.truncation is not None:
+        keep = tok.truncation
+        tok.no_truncation()
+        yield tok
+        tok.enable_truncation(**keep)
+    else:
+        yield tok
+
+
+def tokenize_text(
+    text,
+    tokenizer=None,
+    model: Optional[NomicTextEmbeddingModel] = None,
+    add_special_tokens=False,
+):
+    if tokenizer is None:
+        if model is None:
+            raise ValueError("Either tokenizer or model must be provided")
+        tokenizer = load_tokenizer(model)
+    # padding and truncation are handled by tokenizer
+
+    all_tokens = tokenizer.encode(text, add_special_tokens=add_special_tokens).ids
+    if len(all_tokens) == 0:
+        logger.warning(f"Zero tokens generated from text.")
+        all_tokens = tokenize_text(EMPTY_PLACEHOLDER, add_special_tokens=False).ids
+    return all_tokens
+
+
+def create_sm_request_for_batch(texts: List[str], tokenizer: Tokenizer):
```

Comment:
nit: it took me a second to figure out what `sm` stands for, could rename to something like `create_sagemaker_request`

Reply:
I almost want to make it hidden so people just use the generator but agreed on rename. thanks!